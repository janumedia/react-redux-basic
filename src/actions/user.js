import { ADD_USER, DELETE_USER } from "../constants/action-type";

export const addUser = (newUser) => ({
    type: ADD_USER,
    payload: newUser
})

export const deleteUser = (userID) => ({
    type: DELETE_USER,
    payload: userID
})
