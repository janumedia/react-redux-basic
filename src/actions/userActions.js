import * as types from "../constants/action-types";

export const addUser = (newUser) => ({
    type: types.ADD_USER,
    payload: newUser
})

export const deleteUser = (userID) => ({
    type: types.DELETE_USER,
    payload: userID
})

export const getUsers = () => ({
    type: types.GET_USERS,
    payload: {url: '/data/users.json'}
})
