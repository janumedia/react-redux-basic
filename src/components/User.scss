.User {
    background-color: #ccc;
    margin-top: 20px;
    transition: opacity 1s;
    opacity: 1;
    position: relative;
    &.hidden {
        opacity: 0;
    }
    &.deleting {
        opacity: 0.4;
    }
    div:first-child {
        p {
            padding: 10px;
            margin: 0;
            &:last-child {
                padding-top: 0;
            }
        }
    }
    div:last-child {
        position: absolute;
        right: 10px;
        top: 0;
        transform: translateY(50%);
        button {
            padding: 10px;
            text-transform: uppercase;
        }
    }
}