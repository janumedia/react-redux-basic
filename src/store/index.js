import { createStore, applyMiddleware } from 'redux';

import userReducer from '../reducers/userReducer';
import userMiddleWare from '../middleware/userMiddleWare';

export default createStore (userReducer, applyMiddleware(userMiddleWare));