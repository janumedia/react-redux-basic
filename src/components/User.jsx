import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import {deleteUser} from '../actions/userActions';

import './User.scss';

const mapDispatchToProps = dispatch => {
    return {
        deleteUser: userID => dispatch(deleteUser(userID))
    };
};

class User extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            loaded: false
        }
    }

    componentDidMount() {
        this.timerID = setTimeout(() => {
            this.setState({loaded: true})
        }, 10)
    }

    componentWillUnmount() {
        clearTimeout(this.timerID);
    }

    //optimization only render new user
    shouldComponentUpdate(nextProps, nextState) {
        return (this.props === nextProps)
    }

    deleteUser(e) {
        this.props.deleteUser(this.props.id);
    }

    render() {
        const {name, role} = this.props;
        return (
            <div className={`User ${this.state.loaded ? '' : 'hidden'}`}>
                <div>
                    <p>Name: {name}</p>
                    <p>Role: {role}</p>
                </div>
                <div>
                    <button onClick={this.deleteUser.bind(this)}>Delete</button>
                </div>
            </div>
        );
    }
}

User.propTypes = {
    name: PropTypes.string,
    role: PropTypes.string
}

export default connect(null, mapDispatchToProps)(User);