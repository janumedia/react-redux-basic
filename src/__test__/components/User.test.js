import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import store from '../../store';
import * as userActions from '../../actions/userActions';
import User from '../../components/User';

jest.useFakeTimers();

describe('User component', () => {

    const props = {
        store,
        name: 'John Doe',
        role: 'CEO',
        id: 1
    }

    const simulatePromiseResolved = () => new Promise(resolve => {
        jest.runOnlyPendingTimers();
        setImmediate(resolve);
    });

    let wrapper, button, deleteUserSpy;

    beforeEach(() => {
        //use .dive() is important for redux-connected component
        //https://github.com/airbnb/enzyme/issues/1002#issuecomment-310796612
        wrapper = shallow(<User {...props}/>).dive();
        button = wrapper.find('button');
        deleteUserSpy = jest.spyOn(userActions, 'deleteUser');
    });

    afterEach(() => {
        deleteUserSpy.mockRestore();
        jest.clearAllTimers();
    });
    
    it('should rendered properly', () => {
        expect(wrapper.type()).toEqual('div');
        expect(wrapper.hasClass('User')).toBe(true);
        expect(wrapper.find('p').first().text()).toEqual(`Name: ${props.name}`);
        expect(wrapper.find('p').last().text()).toEqual(`Role: ${props.role}`);
        expect(wrapper.find('button')).toHaveLength(1);
        expect(wrapper.state('loaded')).toBe(false);
        expect(wrapper.hasClass('hidden')).toBe(true);
        expect(toJson(wrapper)).toMatchSnapshot();
        //process pending timer
        jest.runOnlyPendingTimers();
        expect(wrapper.state('loaded')).toBe(true);
        //update component
        wrapper.update();
        expect(wrapper.hasClass('hidden')).toBe(false);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should triggered "deleteUser" on button clicked', () => {
        button.simulate('click');
        expect(deleteUserSpy).toHaveBeenCalled();
    });

    it('should triggered "deleteUser" only once upon continous button clicked', () => {
        button.simulate('click');
        button.simulate('click');
        button.simulate('click');
        expect(deleteUserSpy).toHaveBeenCalledTimes(1);
    });

    it('should update state on button clicked and "deleteuser" resolved', async() => {
        button.simulate('click');
        expect(wrapper.state('deleting')).toBe(true);
        
        await simulatePromiseResolved();

        expect(wrapper.state('deleting')).toBe(false);
    });

    it('should able button click after "deleteuser" resolved', async() => {
        button.simulate('click');
        expect(deleteUserSpy).toHaveBeenCalledTimes(1);
        expect(wrapper.state('deleting')).toBe(true);
        
        button.simulate('click');
        //until resolved action will not able to call
        expect(deleteUserSpy).toHaveBeenCalledTimes(1);

        await simulatePromiseResolved();

        expect(wrapper.state('deleting')).toBe(false);
        button.simulate('click');
        expect(deleteUserSpy).toHaveBeenCalledTimes(2);
        expect(wrapper.state('deleting')).toBe(true);
    });

    it('should should triggered "deleteUser" and unmount properly', async () => {
        button.simulate('click');
        
        expect(toJson(wrapper)).toMatchSnapshot();
        
        //simulate unmount
        wrapper.unmount();
        //update component
        wrapper.update();
        
        expect(deleteUserSpy).toHaveBeenCalledTimes(1);
        
        await simulatePromiseResolved();

        expect(toJson(wrapper)).toMatchSnapshot();
    });

});