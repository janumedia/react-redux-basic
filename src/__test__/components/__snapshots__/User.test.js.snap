// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User component should rendered properly 1`] = `
<div
  className="User hidden "
>
  <div>
    <p>
      Name: 
      John Doe
    </p>
    <p>
      Role: 
      CEO
    </p>
  </div>
  <div>
    <button
      onClick={[Function]}
    >
      Delete
    </button>
  </div>
</div>
`;

exports[`User component should rendered properly 2`] = `
<div
  className="User  "
>
  <div>
    <p>
      Name: 
      John Doe
    </p>
    <p>
      Role: 
      CEO
    </p>
  </div>
  <div>
    <button
      onClick={[Function]}
    >
      Delete
    </button>
  </div>
</div>
`;

exports[`User component should should triggered "deleteUser" and unmount properly 1`] = `
<div
  className="User hidden deleting"
>
  <div>
    <p>
      Name: 
      John Doe
    </p>
    <p>
      Role: 
      CEO
    </p>
  </div>
  <div>
    <button
      onClick={[Function]}
    >
      Deleting
    </button>
  </div>
</div>
`;

exports[`User component should should triggered "deleteUser" and unmount properly 2`] = `""`;
